Src/Base/AMReX_TypeTraits.H

```diff
@@ -77,23 +77,47 @@ namespace amrex
     struct IsParticleContainer : public std::is_base_of<ParticleContainerBase, T>::type {};
 #endif
     
-#if defined(AMREX_USE_GPU)
+#if defined(AMREX_USE_CUDA)
 
     template <class T, class Enable = void>
-    struct IsDeviceLambda : std::false_type {};
+    struct HostRunnable : std::true_type {};
+    //
+    template <class T>
+    struct HostRunnable<T, amrex::EnableIf_t<__nv_is_extended_device_lambda_closure_type(T)> >
+        : std::false_type {};
 
     template <class T, class Enable = void>
-    struct IsHostDeviceLambda : std::false_type {};
-
-#if defined(AMREX_USE_CUDA)
+    struct DeviceRunnable : std::false_type {};
+    //
     template <class T>
-    struct IsDeviceLambda<T, amrex::EnableIf_t<__nv_is_extended_device_lambda_closure_type(T)> >
+    struct DeviceRunnable<T, amrex::EnableIf_t<__nv_is_extended_host_device_lambda_closure_type(T)
+                                            or __nv_is_extended_device_lambda_closure_type(T)> >
         : std::true_type {};
 
+    template <class T, class Enable = void>
+    struct HostDeviceRunnable : std::false_type {};
+    //
     template <class T>
-    struct IsHostDeviceLambda<T, amrex::EnableIf_t<__nv_is_extended_host_device_lambda_closure_type(T)> >
+    struct HostDeviceRunnable<T, amrex::EnableIf_t<HostRunnable<T>::value
+                                                   and DeviceRunnable<T>::value> >
         : std::true_type {};
-#endif
+
+#elif defined(AMREX_USE_HIP)
+
+    // xxxxx HIP todo
+    template <class T, class Enable = void>
+    struct HostRunnable : std::true_type {};
+
+    template <class T, class Enable = void>
+    struct DeviceRunnable : std::true_type {};
```

Comment:
Won't this break compilation for HIP and DPC++ since e.g. HostDeviceParallelFor now uses HostDeviceRunnable but that is only defined for CUDA?

Reply:
Yes, you are right.  I have update the PR moving `HostDeviceRunnable` out of `AMREX_USE_CUDA` into `AMREX_USE_GPU` block.